### PR TITLE
EditPersonDescriptor: update methods to perform defensive copy

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
```
getTags() returns the same set of tags stored by this object, while
setTags(Set<Tag>) assigns the given Set<Tag> to its field without
defensive copy.

Callers can modify the state of the Set<Tag> externally, thus breaking
EditPersonDescriptor’s encapsulation.

Let's update getTags() to return an unmodifiable set, and 
setTags(Set<Tag>) to perform a defensive copy of the input before 
assigning it to its field.
```

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 1]
```diff
@@ -136,11 +137,11 @@ public boolean equals(Object other) {
...
         public EditPersonDescriptor(EditPersonDescriptor toCopy) {
-            this.name = toCopy.name;
-            this.phone = toCopy.phone;
-            this.email = toCopy.email;
-            this.address = toCopy.address;
-            this.tags = toCopy.tags;
+            setName(toCopy.name);
+            setPhone(toCopy.phone);
+            setEmail(toCopy.email);
+            setAddress(toCopy.address);
```
Comment (yamgent) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Any rationale behind changing these to call `set...()`? Also seems to be better to be a separate commit on its own since this commit is about `setTags()`. :P

Comment (Zhiyuan-Amos) [reply src/main/java/seedu/address/logic/commands/EditCommand.java]: It's just for consistency, otherwise we'll have `setTags(toCopy.tags)`, but the rest will remain as `this.name = toCopy.name` ><

Which do you think is better?

Comment (yamgent) [reply src/main/java/seedu/address/logic/commands/EditCommand.java]: Using `set...()` for consistency is better, since it allows us to future proof them in case the implementation for them changes (e.g. if `Name` becomes no longer mutable, means that we will have to change here to do defensive copy, etc...).

Comment (Zhiyuan-Amos) [reply src/main/java/seedu/address/logic/commands/EditCommand.java]: ok got it! I'll do it in a separate commit. ^^

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 2]
```diff
@@ -183,11 +185,11 @@ public void setAddress(Address address) {
         }
 
         public void setTags(Set<Tag> tags) {
-            this.tags = tags;
+            this.tags = (tags != null) ? new HashSet<>(tags) : null;
         }
 
         public Optional<Set<Tag>> getTags() {
-            return Optional.ofNullable(tags);
+            return (tags != null) ? Optional.of(Collections.unmodifiableSet(tags)) : Optional.empty();
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: @yamgent do you think this parentheses around `tags != null` is necessary? I recall getting corrected by Paul sometime earlier before :P The parentheses for `this.tags = (tags != null) ? new HashSet<>(tags) : null;` few lines above is necessary though right? Otherwise looks quite unreadable imo.

Comment (yamgent) [reply src/main/java/seedu/address/logic/commands/EditCommand.java]: > do you think this parentheses around tags != null is necessary?

I don't mind having parentheses around. Since adding parentheses doesn't impede readability, it's fine.

> The parentheses for this.tags = (tags != null) ? new HashSet<>(tags) : null; few lines above is necessary though right? Otherwise looks quite unreadable imo.

Yes.

**Comment by damithc** [comment]
[1/3]
1. commit message: `setTags(Set<Tag>) do not perform ` --> `does not`?
2. commit message title: don't we usually use a lower case letter after the colon?
3. Should we mention the defensive copying in the header comment?

[2/3]
* points 2 and 3 above

[3/3]
* point 2 above

**Comment by damithc** [comment]
The first sentence of the javadoc should be an overview of the method, rather than a mention of its peculiarities :-)

**Comment by Zhiyuan-Amos** [comment]
> The first sentence of the javadoc should be an overview of the method, rather than a mention of its peculiarities :-)

Can't think of a way to phrase it well >< `Sets a copy of {@code tags}?`

**Comment by damithc** [comment]
> Can't think of a way to phrase it well >< Sets a copy of {@code tags}?

Yes, something like that. In cases where the peculiarities needs more explanations, they can be given in the subsequent sentences.
```java
/**
 * Sets the ... of the ... object to given {@code ...}.  A defensive copy of ... is used internally.
 */

**Comment by damithc** [comment]
Looks good. On a related note, when a class doesn't have any set methods, how safe is it to say 'the objects are immutable and therefore no need for defensive copying'? Isn't it possible that setters  are added in future, because when someone looks at the class there is no indication that the class is to remain immutable?

**Comment by Zhiyuan-Amos** [comment]
> Looks good.

Does this mean that I can merge? XD

> Isn't it possible that setters are added in future, because when someone looks at the class there is no indication that the class is to remain immutable?

In our classes we leave header comments to denote that classes are immutable. For e.g. [Name](https://github.com/se-edu/addressbook-level4/blob/master/src/main/java/seedu/address/model/person/Name.java#L8). According to [this answer on SO](https://softwareengineering.stackexchange.com/a/272229), it seems that doing this is sufficient.

We can annotate `@Immutable` in our code base too (import stmt: `import jdk.nashorn.internal.ir.annotations.Immutable;`), but it doesn't do any compilation check to make sure that this class is indeed immutable. I think [Checker Framework](https://checkerframework.org/manual/checker-framework-manual.pdf) may be of some use, but have to take a deeper look into it.

**Comment by damithc** [comment]
> Does this mean that I can merge? XD

Yes, if all relevant immutable classes have such header comments, that should be ok for now.
